Feature/simple ddos detection

mkacper opened the pull request.

mkacper committed:
Add tests for Policymaker and PolicymakerStash modules

  - test if Policymaker.forward?/1 correctly detect DDoS and
    behave as we expect when the attack was detected
  - test if PolicymakerStash starts

mkacper committed:
Add feature which enables very simple DDoS detection.

  - Write Policymaker and PolicymakerStash moudles. The modules
    provide very primitive algorithm for DDoS detection.
  - Modify Forwarder module to implement the algorithm.
  - Modify the main supervisor to run PolicymakerStash under it.
  - Add neccessary configuration.

mkacper committed:
Improve Policymaker test module

Use config values instead of hardcode them.

mkacper committed:
Modify "send message to controller" test from SdnEpc.ForwarderTest

Divide the test into two cases: one for normal network traffic,
the other for DDoS. Change async option in the module to false.

mkacper commented on lib/sdn_epc/forwarder.ex:
```diff
@@ -88,20 +89,21 @@ defmodule SdnEpc.Forwarder do
-  def handle_cast({:save_datapath_id, datapath_id}, _state) do
-    {:noreply, %{datapath_id: datapath_id}}
+  def handle_cast({:save_datapath_id, datapath_id}, state) do
+    {:noreply, Map.put(state, :datapath_id, datapath_id)}
```
GenServer state should be a struct.

mkacper commented on lib/sdn_epc/policymaker_stash.ex:
```diff
@@ -0,0 +1,14 @@
+defmodule SdnEpc.PolicymakerStash do
```
State from this module should be kept in Forwarder process.

mkacper commented on lib/sdn_epc/forwarder.ex:
```diff
@@ -78,7 +78,8 @@ defmodule SdnEpc.Forwarder do
   def init(:ok) do
-    {:ok, %{}}
+    {:ok, %{ofp_messages_stats: %{start_time: System.system_time(:seconds),
+                                  count: 0}}}
```
Policymaker module should initialize ofp_messages_stats struct.

mkacper commented on lib/sdn_epc/forwarder.ex:
```diff
@@ -88,20 +89,21 @@ defmodule SdnEpc.Forwarder do
-  def handle_cast({:send_msg_to_controller, switch_id, msg}, state) do
-    msg_converted = SdnEpc.Converter.convert(msg)
-    @ofp_channel.send(switch_id, msg_converted)
+  def handle_cast({:send_msg_to_controller, switch_id, msg},
```
Forwarder can be blocking/forwarding state.

mkacper commented on lib/sdn_epc/forwarder.ex:
```diff
@@ -113,4 +115,17 @@ defmodule SdnEpc.Forwarder do
   end
+
+  # Helper functions
+
```
Decision if to drop/forware is to be made on the handle_cast level.

mkacper commented on lib/sdn_epc/policymaker.ex:
```diff
@@ -0,0 +1,46 @@
+  # API functions
+
+  @spec forward?(%{start_time: integer(),
+                   count: integer()}) :: boolean()
```
use `non_neg_integer` type

mkacper commented on test/sdn_epc/forwarder_test.exs:
```diff
@@ -26,17 +29,41 @@ defmodule SdnEpc.ForwarderTest do
 
     # WHEN
+    SdnEpc.PolicymakerStash.set_drop_start_time(nil)
+    SdnEpc.Forwarder.send_msg_to_controller({'1', nil}, msg)
```
Use `setup`

mkacper commented on test/sdn_epc/forwarder_test.exs:
```diff
@@ -26,17 +29,41 @@ defmodule SdnEpc.ForwarderTest do
+    # GIVEN
+    datapath_id = '00:00:00:00:00:00:00:01'
+    msg = OfpMessage.get(:packet_in)
+    timeout =
```
Put that into a well named function

mkacper commented on test/sdn_epc/forwarder_test.exs:
```diff
@@ -26,17 +29,41 @@ defmodule SdnEpc.ForwarderTest do
+    # WHEN
+    SdnEpc.PolicymakerStash.set_drop_start_time(nil)
+    for _ <- 1..20 do
+      SdnEpc.Forwarder.send_msg_to_controller({'00', self()}, msg)
```
'00' -> make_random_ofp_mesage

mkacper commented on test/sdn_epc/policymaker_test.exs:
```diff
@@ -0,0 +1,35 @@
+    # WHEN
+    SdnEpc.PolicymakerStash.set_drop_start_time(nil)
+    timer =
+      fn() -> Process.send_after(self, :ok, @drop_duration * 1000 + 1000) end
```
put the time calculactions into a function

mkacper commented on test/sdn_epc/policymaker_test.exs:
```diff
@@ -0,0 +1,35 @@
+
+  defp check_forward_policy(msgs_stats) do
+    for _ <- 1..100 do
+      assert !SdnEpc.Policymaker.forward?(msgs_stats)
```
refute

mkacper replied:
one msg engough

mkacper commented on test/sdn_epc/policymaker_test.exs:
```diff
@@ -0,0 +1,35 @@
+    end
+    receive do
+      :ok ->
+        msgs_stats = %{start_time: System.system_time(:seconds), count: 0}
```
move timer to the afer clause of recevive

mkacper committed:
Change Forwarder way of working in terms of DDoS threat

  - add mode of working instead of call Policymaker.forward?/1
    every single time when message arrives
  - cosmetic changes: use struct instead of pure map for keeping
    GenServer state
  - modify Forwarder tests accordingly to changes in Forwarder

mkacper committed:
Write tests for SdnEpc.Policymaker module

  - test if ddos is properly detecting
  - test if Forwarder is being killed when dosen't respond to
    switch mode call
  - add needed configuartion
  - move calculate_receive_timeout/0 to TestHelper module as it
    became common (modify the function to take one parameter)

mentels commented on lib/sdn_epc/forwarder.ex:
```diff
@@ -75,11 +77,19 @@ defmodule SdnEpc.Forwarder do
+  and :blocking.
+  """
+  @spec switch_mode(mode :: atom()) :: term()
+  def switch_mode(mode) do
```
I don't think exposing the mode out of the forwarder is a good idea. I suggest `Forwarder.block()` API.

mkacper committed:
Change Forwarder switch mode of working API

  - add SdnEpc.Forwarder.block/0 and SdnEpc.Forwarder.forward/0
  - write tests for newly added functions
  - join two test into one in Forwarder test module as it is
    enaugh
  - correct a typo in Forwarder module

mkacper committed:
Fix PolicymakerTest `detect DDoS` test

  - send real ofp message to controller instead of fake one

mkacper committed:
Change whole Policymaker module

  - change ddos detection algorithm
  - modify Forwarder change mode of working API - add timeout to
    GenServer calls to make possible restart the Forwarder when
    dosen't respond
  - write apropriate tests
  - delete relevant PolicymakerStash module
  - code style changes in configuration file

mkacper commented:
@mentels, do you have any idea how to reach 100% test coverage of Policymaker module? Test coverage of this module is about 92% now and I don't have any idea what tests are missing. [Here](https://github.com/mkacper/sdn_epc/pull/6/files#diff-ea61f980f14b86d3b05f8cf2160ea521) you can see the PolicymakerTest module.

mkacper commented on test/sdn_epc/policymaker_test.exs:
```diff
@@ -0,0 +1,82 @@
+    assert_receive(^datapath_id)
+  end
+
+  test "kill Forwarder after switch mode call timeout" do
```
@mentels, Could you help me with that test? It dosen't pass. Could you give me some advice how to write it properly?

mkacper committed:
Fix ebert issues from PR #6

mentels commented on lib/sdn_epc/policymaker.ex:
```diff
@@ -0,0 +1,101 @@
+  @msgs_limit Application.get_env(:sdn_epc, :pic_msgs_limit)
+  @drop_duration Application.get_env(:sdn_epc, :pic_drop_duration)
+  @check_policy_timeout 1000
+  @moduledoc false
```
Put the `@moduledoc` at the top of the module.

mentels commented on test/sdn_epc/policymaker_test.exs:
```diff
@@ -0,0 +1,82 @@
+    fill_up_forwarder_when_count_duration(sw_mode_timeout, send_fake_msg)
+
+    # THEN
+    refute Process.whereis(SdnEpc.Forwarder)
```
Forwarder is being restarted by the supervisor?

mkacper replied:
Should be. I want to test if exit signal is send to the Forwarder after switch mode timeout is exceeded when Policymaker try to change Forwarder mode of working.

mentels commented:
Use a monitor to get a notification of the process going down.

On Mon, Apr 24, 2017 at 11:35 AM, Kacper Mentel <notifications@github.com>
wrote:

> *@mkacper* commented on this pull request.
> ------------------------------
>
> In test/sdn_epc/policymaker_test.exs
> <https://github.com/mkacper/sdn_epc/pull/6#discussion_r112902649>:
>
> > +    msg = OfpMessage.get(:packet_in)
> +    datapath_id = '01'
> +    send_fake_msg =
> +    fn() ->
> +      for _ <- 1..100 do
> +        SdnEpc.Forwarder.send_msg_to_controller(datapath_id, msg)
> +      end
> +    end
> +
> +    # WHEN
> +    Application.put_env(:sdn_epc, :pm_sw_mode_timeout, sw_mode_timeout)
> +    Process.send_after(self(), :stop, count_duration_timeout)
> +    fill_up_forwarder_when_count_duration(sw_mode_timeout, send_fake_msg)
> +
> +    # THEN
> +    refute Process.whereis(SdnEpc.Forwarder)
>
> Should be. I want to test if exit signal is send to the Forwarder after
> switch mode timeout is exceeded when Policymaker try to change Forwarder
> mode of working.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mkacper/sdn_epc/pull/6#discussion_r112902649>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAtsQJpjCZFzRScu5DD3_4sVH0tZ1CaUks5rzGzXgaJpZM4M-XGV>
> .
>



-- 
Szymon Mentel
Erlang Solutions Ltd.
